### PR TITLE
Merges #12 and #13 into `main`

### PR DESCRIPTION
**Pull request by gynt**
Merges #12 and #13 into `main`

Testing is required though. Can you take care of that @Krarilotus ?

### CODE REVIEW

**Comment by Krarilotus** [comment]
@gynt sorry, what's this abt? Did xou just merge my chnages?

**Comment by gynt** [comment]
> @gynt sorry, what's this abt? Did xou just merge my chnages?

I merged the two PR's we created into a development branch. I wanted to ask you if you can test this new development branch whether everything works as expected.

https://github.com/UnofficialCrusaderPatch/extension-aicloader/archive/refs/heads/main.zip

**Comment by Krarilotus** [comment]
> I merged the two PR's we created into a development branch. I wanted to ask you if you can test this new development branch whether everything works as expected.
> 
> https://github.com/UnofficialCrusaderPatch/extension-aicloader/archive/refs/heads/main.zip

So does that mean, that the AIC value fields behave again differently?
Like the AIC Value for the DefSiegeEngineGoldThreshold seems reverted but no changelog was added again, so i am not sure if people would expect this.

So does it 'work as expected' Not for me, as i think now i would expect it to work as before, but thats mainly a news/documentation issue. Maybe we could add the feautre of how the DefSiegeEngineGoldThreshold back in and also merge the news feature, announcing that it now works different than before.

But on the other hand we should also make sure, that the AI-Swapper has a version change that reflects the incompatibility with previous versions, as its not compatible to use this in multiplayer with a previous version with AI if they have differently working AIC values.

Anyways, i think we can merge everything into main, once we clear up confusions. Because the fix for the DefSiegeEngineGoldThreshold  was to make it work as we expect it from its name and what we think that firefly intended, its not a new AIC value.
Hence we should with this change, make it possible that all plugins that rely on it get a new AIC value added automatically, if they were using the old version. Which should be handled by the incompatibility thing. I would propose a name like 'oldInvertedDefSiegeEngineGoldThreshold'.

But: To reduce maintanance, maybe its more feasable to instead rename the value entirely, and make it clear what it does while also not sounding confusing, then make sure the old value is correctly handled and reported as AI is cheating.

So TL:DR
- find a good solution for the DefSiegeEngineGoldThreshold 
- reduce maintanance
- don't make incompatibility changes without changing versioning accordingly
- don'T invalidate all existing AICs by this change, but handle them proper and report AIs cheating

**Comment by gynt** [comment]
The plan which this PR addresses is to revert things so it works like the old ucp. This plan was agreed on in our voice chat and here: https://github.com/UnofficialCrusaderPatch/extension-aicloader/issues/10#issuecomment-2373962201

The goal was to make hackerman ai work again.

So you would like to see a changelog added in the description MD file before merge? That makes sense

I don't see the need for a version change of the ai-Swapper if the AIC loader already changes/changed versions. If developers of ais want to make use of new AIC logic they should add an explicit dependency on the right AIC loader versions.

Multiplayer compatibility is being addressed with the config verification protocol which is still in the pipeline. 

So I looked at the code for this threshold. The way it works is that the value is added to the current gold of the player, if that sum reaches the required gold value, then siege units are created. I don't see anything wrong with that per se. It does mean it can cheat with negative gold, but as far as I can tell it is by design. The threshold is reached quicker if you set a positive value and you can set a required gold reserve if you set a negative value. e.g. if you set -200 gold you slow down siege engine production because 200 + the cost of the siege engine is required.

I think threshold is kind of the wrong word indeed if you think of it as "increasing the threshold (making something more difficult)", but if you think of it as "manipulating the threshold" it is fine, then higher values make sense because you enhance or speed up creation. So it is semantics and can be solved with a proper explanation, right? Is the issue that no one reads explanations?

I think for a new AIC value that is reversed "Reserve" would be a good name. Shall I change the name to that in this PR?

---

Does the new AIC value raise a warning? I don't think so, as only when a AIC config tries to use non existing AIC field you get a warning.

---

TLDR
- rename the new (reversed) AIC field to "DefSiegeEngineGoldReserve"
- add changelog to the description.md or a separate changelog md is hosted that is linked in the description.

**Comment by Krarilotus** [comment]
> I think for a new AIC value that is reversed "Reserve" would be a good name. Shall I change the name to that in this PR?
Yes please change the name to something that makes actual sense. So by design the AI can basically cheat siege engines in the vanilla game by making this value X, where the total sum of cost for siege engines cheated is X. At least that's what we experienced. Reserve would indeed sound like something else though, it sounds like the AI reserves this amount of gold for making cheat engines, which is not really correct, if it can just go in the negatives, for X amount to make siege engines but the negative then is reset to 0.

So the proper 'fix' in this sense would probably not be touching the vanilla behaviour for checking this value, nor name this value Threshold then, but instead make sure, the AI can only spend gold on siege engines, that it actually possess. Same as it also cheats the resources it would need to set up a tent. So for example, if the AI does not possess pitch, but wants to build a fire ballista that would cost pitch in the rebalance, it can do so without paying any pitch, if it doesn't have pitch. I am not even sure if it tries to buy it, if it could. But is this desired?

What is the purpose of this value then? I think it serves the AI for cheating engines as you pointed out on purpose, by letting it have a higher threshold of magically generating gold out of thin air.
-> In this sense mabye the value would better be named: defSiegeEngineCostReduction, and be added as the new default while preserving the old value name as an alternative for those that got used to it...

Also on a note: defensive siege engines for AIs are severely broken in multiple ways:
- they get cheated based on this discussed value
- siege engine tents in general are offset due to them being placed by their corner piece instead of their middle piece, having AIs sometimes bug out setting siege engines based on map rotation
- If too many defensive siege engines are placed in the AIV, some will never be built, and some will just be placed as tents but also never be manned

And then there are many more bugs around siege engines for offensive and harassing:
- they can only ever build them in the same spot for harassing siege engines
- they destroy their own siege engines sometimes randomly by overbuilding them

____________________________________________
You are right, we agreed on that it should reflect old UCP behaviour, but:
- add some not to description
- make new name default: defSiegeEngineCostReduction
